tests/functional/utils/convergence_analysis.py

```diff
@@ -0,0 +1,313 @@
+"""Module containing a class for performing spatio-temporal convergence analysis."""
+from __future__ import annotations
+
+import warnings
+from copy import deepcopy
+from typing import Optional, Union
+
+import numpy as np
+from scipy import stats
+
+import porepy as pp
+
+
+class ConvergenceAnalysis:
+    """Performing a convergence analysis on a model.
+
+    Current support:
+        - Simplicial and Cartesian grids in 2d and 3d. Tensor grids are not supported.
+        - Static and time-dependent models.
+        - If the model is time-dependent, we assume that the ``TimeManager`` is
+          instantiated with a constant time step.
+
+    Parameters:
+        model_class: Model class for which the analysis will be performed.
+        model_params: Model parameters. We assume that it contains all parameters
+            necessary to set up a valid instance of :data:`model_class`.
+        levels: ``default=1``
+
+            Number of refinement levels associated to the convergence analysis.
+        in_space: ``default=True``
+
+            Whether convergence in space should be performed.
+        in_time:  ``default=False``
+
+            Whether convergence in time should be performed.
+        spatial_rate: ``default=1``
+
+            Rate at which the mesh size(s) should be refined. For example, use
+            ``spatial_rate=2`` for halving the mesh size(s) in-between levels.
+        temporal_rate: ``default=1``
+
+            Rate at which the time step size should be refined. For example,
+            use ``temporal_rate=2`` for halving the time step size in-between levels.
+
+    """
+
+    def __init__(
+        self,
+        model_class,
+        model_params: dict,
+        levels: int = 1,
+        in_space: bool = True,
+        in_time: bool = False,
+        spatial_rate: int = 1,
```

Comment:
Perhaps spatial/temporal_refinement_rate, to distinguish from convergence rates?

Reply:
I agree with Ivar's suggestion, for clarity insert refinement before rate.

Reply:
Suggestion incorporated.